### PR TITLE
Added hlo/lhlo emitters for Abs, Ceil, Convert, Cos, Negate, Remainder, Sign and Tanh ops.

### PR DESCRIPTION
**Pull request by dfki-jugr**

### CODE REVIEW

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 1]
```diff
@@ -393,5 +393,118 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+TEST_F(LhloGenTest, Abs) {
+  CompileAndVerifyIr(R"(
+HloModule Abs
+ENTRY %Abs (val: f32[2,2]) -> f32[2,2] {
+  %val = f32[2,2]{1,0} parameter(0)
+  ROOT %abs = f32[2,2]{1,0} abs(f32[2,2]{1,0} %val)
+})",
+                     R"(
+;CHECK: func @abs(%[[ARG0:.*]]: [[TYPE:.*]], %[[ARG1:.*]]: [[TYPE]]) {
+;CHECK:   "xla_lhlo.abs"(%[[ARG0]], %[[ARG1]]) {name = "abs"} : ([[TYPE]], [[TYPE]]) -> ()
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: I think we removed the preservation of the name attribute a while ago. Have you synced recently? Please remove it from tests.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 2]
```diff
@@ -393,5 +393,118 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+TEST_F(LhloGenTest, Ceil) {
+  CompileAndVerifyIr(R"(
+HloModule Ceil
+ENTRY %Ceil (val: f32[2,2]) -> f32[2,2] {
+  %val = f32[2,2]{1,0} parameter(0)
+  ROOT %ceil = f32[2,2]{1,0} ceil(f32[2,2]{1,0} %val)
+})",
+                     R"(
+;CHECK: func @ceil(%[[ARG0:.*]]: [[TYPE:.*]], %[[ARG1:.*]]: [[TYPE]]) {
+;CHECK:   "xla_lhlo.ceil"(%[[ARG0]], %[[ARG1]]) {name = "ceil"} : ([[TYPE]], [[TYPE]]) -> ()
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: Same here

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 3]
```diff
@@ -393,5 +393,118 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+;CHECK: }
+      )");
+}
+
+TEST_F(LhloGenTest, Convert) {
+  CompileAndVerifyIr(R"(
+HloModule Convert
+ENTRY %Convert (val: f32[2,2]) -> f32[2,2] {
+  %val = f32[2,2]{1,0} parameter(0)
+  ROOT %convert = f32[2,2]{1,0} convert(f32[2,2]{1,0} %val)
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: Can you use different types here? Using the same type is the least interesting case.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 4]
```diff
@@ -393,5 +393,118 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+  CompileAndVerifyIr(R"(
+HloModule Rem
+ENTRY %Rem(x: f32[2,2], y: f32[2,2]) -> f32[2,2] {
+  %x = f32[2,2]{1,0} parameter(0)
+  %y = f32[2,2]{1,0} parameter(1)
+  ROOT %rem = f32[2,2]{1,0} remainder(f32[2,2]{1,0} %x, f32[2,2]{1,0} %y)
+})",
+                     R"(
+;CHECK: func @remainder(%[[ARG0:.*]]: [[TYPE:.*]], %[[ARG1:.*]]: [[TYPE]], %[[ARG2:.*]]: [[TYPE]]) {
+;CHECK:   "xla_lhlo.remainder(%{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %{{.*}}, %[[ARG0]], %[[ARG1]], %[[ARG2]]
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: Why is this pattern so complex? I would have expected that remainder only uses 3 arguments.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 5]
```diff
@@ -393,5 +393,118 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+TEST_F(LhloGenTest, Cos) {
+  CompileAndVerifyIr(R"(
+HloModule Cos
+ENTRY %Cos (val: f32[2,2]) -> f32[2,2] {
+  %val = f32[2,2]{1,0} parameter(0)
+  ROOT %cos = f32[2,2]{1,0} cos(f32[2,2]{1,0} %val)
+})",
+                     R"(
+;CHECK: func @cos(%[[ARG0:.*]]: [[TYPE:.*]], %[[ARG1:.*]]: [[TYPE]]) {
+;CHECK:   "xla_lhlo.cos"(%[[ARG0]], %[[ARG1]]) {name = "cos"} : ([[TYPE]], [[TYPE]]) -> ()
```
Comment (sherhut) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: Name attribute here and below.

**Review by sherhut — CHANGES_REQUESTED** [review]
Mega-nit: Use present tense in commit messages. ("Add missing...").

**Comment by sherhut** [comment]
Thanks for fixing the tests. This looks good to go now.

**Review by sherhut — APPROVED** [review]

**Review by sherhut — CHANGES_REQUESTED** [review]
The changes made the test fail.

[  FAILED  ] LhloGenTest.Convert
[  FAILED  ] LhloGenTest.Cos
[  FAILED  ] LhloGenTest.Neg
[  FAILED  ] LhloGenTest.Rem

Seems the CI run from github skips this test. We need to fix that. In the meanwhile, please run the test locally and fix.

**Comment by akuegel** [comment]
The convert op isn't specified correctly in the dialect, it currently expects to have the same input and output shape. But the whole point of the convert op is to be able to convert the *types* of shapes.
The tests generated this error:
FAILED: 'xla_lhlo.convert' op requires all operands to have the same type
We will probably merge this request with a few fixes, but without the convert op.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 6]
```diff
@@ -393,5 +393,162 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+;CHECK: }
+      )");
+}
+
+TEST_F(LhloGenTest, Cos) {
+  CompileAndVerifyIr(R"(
+HloModule Cos
+ENTRY %Cos (val: f32[2,2]) -> f32[2,2] {
+  %val = f32[2,2]{1,0} parameter(0)
+  ROOT %cos = f32[2,2]{1,0} cos(f32[2,2]{1,0} %val)
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: HLO for cos is "cosine"
So needs to be cosine(f32[2,2]{1,0} %val)

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 7]
```diff
@@ -393,5 +393,162 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+})",
+                     R"(
+;CHECK: func @cos(%[[ARG0:.*]]: [[TYPE:.*]], %[[ARG1:.*]]: [[TYPE]]) {
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: When you fix the HLO above, you need to name this cosine as well.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 8]
```diff
@@ -393,5 +393,162 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+;CHECK: }
+      )");
+}
+
+TEST_F(LhloGenTest, Neg) {
+  CompileAndVerifyIr(R"(
+HloModule Neg
+ENTRY %Neg (val: f32[2,2]) -> f32[2,2] {
+  %val = f32[2,2]{1,0} parameter(0)
+  ROOT %neg = f32[2,2]{1,0} neg(f32[2,2]{1,0} %val)
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: neg -> negate for the HLO op.
Similar to the cosine fix, you will also have to adjust func below.

File: tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc [hunk 9]
```diff
@@ -393,5 +393,162 @@ ENTRY %AddReduce (x: f32[100,10], c: f32[]) -> f32[100] {
...
+  CompileAndVerifyIr(R"(
+HloModule Rem
+ENTRY %Rem(x: f32[2,2], y: f32[2,2]) -> f32[2,2] {
+  %x = f32[2,2]{1,0} parameter(0)
+  %y = f32[2,2]{1,0} parameter(1)
+  ROOT %rem = f32[2,2]{1,0} remainder(f32[2,2]{1,0} %x, f32[2,2]{1,0} %y)
+})",
+                     R"(
+;CHECK: func @remainder(%[[ARG0:.*]]: [[TYPE:.*]], %[[ARG1:.*]]: [[TYPE]], %[[ARG2:.*]]: [[TYPE]]) {
+;CHECK:   "xla_lhlo.remainder(%[[ARG0]], %[[ARG1]], %[[ARG2]]) : ([[TYPE]], [[TYPE]], [[TYPE]]) -> ()
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/mlir_gpu/tests/mlir_gpu_lhlo_gen_test.cc]: missing " after remainder:
"xla_lhlo.remainder"

**Review by akuegel — COMMENTED** [review]
Just some comments to show what fixes we had to do to make it work.
In general, you can check https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/hlo_opcode.h
how the HLO ops are written in HLO text.

**Pull request merged**